test/zdtm/static/maps11.c

```diff
@@ -0,0 +1,178 @@
+#include <stdint.h>
+#include <signal.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include "zdtmtst.h"
+
+#ifndef MAP_DROPPABLE
+#define MAP_DROPPABLE 0x08
+#endif
+
+#ifndef MADV_WIPEONFORK
+#define MADV_WIPEONFORK 18
+#endif
+
+const char *test_doc = "Test MAP_DROPPABLE/MADV_WIPEONFORK mappings with 2 processes";
+const char *test_author = "Alexander Mikhalitsyn <aleksandr.mikhalitsyn@canonical.com>";
+
+bool mem_is_zero(const uint8_t *buffer, size_t length)
+{
+	size_t i;
+
+	for (i = 0; i < length; i++)
+		if (buffer[i] != 0)
+			return false;
+
+	return true;
+}
+
+int main(int argc, char **argv)
+{
+	uint8_t *p1, *p2;
+	pid_t pid;
+	int status;
+	const char data[] = "MADV_WIPEONFORK vma data";
+	bool criu_was_there = false;
+	struct stat st1, st2;
+
+	test_init(argc, argv);
+
+	p1 = mmap(NULL, sizeof(data), PROT_READ | PROT_WRITE, MAP_DROPPABLE | MAP_ANONYMOUS, 0, 0);
+	if (p1 == MAP_FAILED) {
+		if (errno == EINVAL) {
+			skip("mmap failed, no kernel support for MAP_DROPPABLE\n");
+			goto skip;
+		} else {
+			pr_perror("mmap failed");
+			return -1;
+		}
+	}
+
+	p2 = mmap(NULL, sizeof(data), PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, 0, 0);
+	if (p2 == MAP_FAILED) {
+		pr_perror("mmap failed");
+		return 1;
+	}
+
+	if (madvise(p2, sizeof(data), MADV_WIPEONFORK)) {
+		pr_perror("madvise failed");
+		return -1;
+	}
+
+	/* contents of this mapping is supposed to be dropped after C/R */
+	memcpy(p1, data, sizeof(data));
+
+	/* contents of this mapping is supposed to be dropped after fork() */
+	memcpy(p2, data, sizeof(data));
+
+	/*
+	 * Let's spawn a process before C/R so our mappings get inherited
+	 * then, after C/R we need to ensure that CRIU memory premapping
+	 * machinery works properly.
+	 *
+	 * It is important, because we restore MADV_WIPEONFORK on a later
+	 * stages (after vma premapping happens) and we need to ensure that CRIU
+	 * handles everything in a right way.
+	 */
+	pid = test_fork();
+	if (pid < 0) {
+		pr_perror("fork failed");
+		return 1;
+	}
+
+	if (pid == 0) {
+		test_waitsig();
+
+		/*
+		 * Both mappings have VM_WIPEONFORK flag set,
+		 * so we expect to have it null-ified after fork().
+		 */
+		if (!mem_is_zero(p1, sizeof(data)) || !mem_is_zero(p2, sizeof(data)))
+			return 1;
+
+		return 0;
+	}
+
+	/*
+	 * A simple way to detect if C/R happened is to compare st_ino
+	 * fields of stat() on the procfs files of the current task.
+	 *
+	 * Hopefully, this terrible hack is never used in real-world applications ;-)
```

Comment:
nit: 80 symbols warp is still preferred...

Reply:
Have fixed!